File: scripts/build_site.py
#!/usr/bin/env python3
"""Assemble the static directory site from site/ plus data/lists.json."""

from __future__ import annotations

import shutil
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SITE = ROOT / "site"
DIST = ROOT / "dist"
LISTS = ROOT / "data" / "lists.json"
HIGHLIGHTS = ROOT / "HIGHLIGHTS.md"


def main() -> None:
    if not SITE.is_dir():
        raise SystemExit(f"Missing site directory: {SITE}")
    if not LISTS.is_file():
        raise SystemExit(f"Missing catalog data: {LISTS}")
    if not HIGHLIGHTS.is_file():
        raise SystemExit(f"Missing highlights: {HIGHLIGHTS}")

    if DIST.exists():
        shutil.rmtree(DIST)
    shutil.copytree(SITE, DIST, ignore=shutil.ignore_patterns(".DS_Store"))

    data_dir = DIST / "data"
    data_dir.mkdir(exist_ok=True)
    shutil.copy2(LISTS, data_dir / "lists.json")
    shutil.copy2(HIGHLIGHTS, DIST / "HIGHLIGHTS.md")
    print(f"Built {DIST} ({LISTS.stat().st_size} bytes of catalog data)")


if __name__ == "__main__":
    main()

File: scripts/build_site.mjs
import { cpSync, mkdirSync, readFileSync, rmSync, statSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const root = join(dirname(fileURLToPath(import.meta.url)), "..");
const site = join(root, "site");
const dist = join(root, "dist");
const listsPath = join(root, "data", "lists.json");
const highlights = join(root, "HIGHLIGHTS.md");

rmSync(dist, { recursive: true, force: true });
cpSync(site, dist, { recursive: true });
mkdirSync(join(dist, "data"), { recursive: true });
cpSync(listsPath, join(dist, "data", "lists.json"));
cpSync(highlights, join(dist, "HIGHLIGHTS.md"));

const template = readFileSync(join(site, "list.html"), "utf8");
const catalog = JSON.parse(readFileSync(listsPath, "utf8"));
if (!Array.isArray(catalog)) {
  throw new Error("Catalog data is not a list");
}
mkdirSync(join(dist, "lists"), { recursive: true });
for (const list of catalog) {
  if (!list?.id) continue;
  const dir = join(dist, "lists", String(list.id));
  mkdirSync(dir, { recursive: true });
  writeFileSync(join(dir, "index.html"), template);
}
rmSync(join(dist, "list.html"), { force: true });

console.log(
  `Built ${dist} (${statSync(listsPath).size} bytes of catalog data, ${catalog.length} interiors)`
);
